`strings::contains()` for multiple scalar search targets

mythrocks opened the pull request:
## Description

This commit adds a new `strings::contains()` overload that allows for the search of multiple scalar search targets in the same call.

The trick here is that a new kernel has been introduced, to extend the "string-per-warp" approach to search for multiple search keys in the same kernel.

This approach allows CUDF to potentially reduce the number of kernels launched for `string::contains()` by a factor of `N`, if all the search keys can be specified in the same call.  This helps reduce the kernel-launch overheads for processes that do large numbers of calls to `string::contains()`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

mythrocks commented:
Tagging @davidwendt.

davidwendt commented on cpp/include/cudf/strings/find.hpp:
```diff
@@ -139,6 +139,31 @@ std::unique_ptr<column> contains(
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return New BOOL8 column
+ */
+std::unique_ptr<table> contains(
+  strings_column_view const& input,
+  std::vector<std::reference_wrapper<string_scalar>> const& targets,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```
I'd prefer this take a `strings_column_view` for the targets and return a list column instead of a table.
This would be more consistent with other multi-target strings APIs.

mythrocks replied:
Appreciated.  That's definitely going to complicate the consumption of the output from `spark-rapids`.

Is there an existing way to transpose a `column<LIST<STRING, N>>` into a `column<STRING>[N]`? (he asked with fading hopefulness.)

davidwendt replied:
I'm not sure I understand your question? What is the `column<LIST<STRING,N>>`?
If you have a list column of strings you can get the child view of the strings column using the `lists_column_view.child()` function.
https://github.com/rapidsai/cudf/blob/feb96cbe39e36d35d673c51270f7316708d24f67/cpp/include/cudf/lists/lists_column_view.hpp#L97

davidwendt commented:
/ok to test

mythrocks commented:
I've placed this in draft, for the moment.  I should do some more testing from `spark-rapids`, to make sure this fits.

hyperbolic2346 commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +373,42 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+  auto const num_targets = d_targets.size();
+  auto const num_rows    = d_strings.size();
+
+  auto const idx = static_cast<size_type>(threadIdx.x + blockIdx.x * blockDim.x);
```
Every time you access `idx` you are extracting the warp, so why add the thread index?

```
auto const warp_idx = static_cast<size_type>(blockIdx.x * blockDim.x);
...
if (warp_idx > num_rows * num_targets) { return; }

auto const lane_idx = thread_idx.x;
auto const str_idx = warp_idx % num_rows;
auto const target_idx = warp_idx / num_rows;
```

hyperbolic2346 replied:
That is a good point. I'm not sure the value of this suggestion.

mythrocks commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +373,42 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+       i += cudf::detail::warp_size) {
+    // check the target matches this part of the d_str data
+    if (d_target.compare(d_str.data() + i, d_target.size_bytes()) == 0) { found = true; }
+  }
```
That makes sense to chase down.

davidwendt commented:
I recommend looking at the changes to `find.cu` here: https://github.com/rapidsai/cudf/pull/15495/files including the commented out code. We could probably remove the fill call.

davidwendt commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +373,42 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+  // each thread of the warp will check just part of the string
+  auto found = false;
+  for (auto i = static_cast<size_type>(idx % cudf::detail::warp_size);
+       !found && ((i + d_target.size_bytes()) <= d_str.size_bytes());
```
The `size_bytes()` is a simple inline member function that returns a single constant member variable.

davidwendt commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +373,42 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+  auto const num_targets = d_targets.size();
+  auto const num_rows    = d_strings.size();
+
+  auto const idx = static_cast<size_type>(threadIdx.x + blockIdx.x * blockDim.x);
```
'idx' is more consistent to the rest of the code in this file

ttnghia replied:
I think it's better to use the common way:
```
auto tid          = cudf::detail::grid_1d::global_thread_id();
auto const stride = cudf::detail::grid_1d::grid_stride();
```

mythrocks commented:
I should have updated here sooner.

It's possible that I was mistaken regarding the improvement from this change. I can't seem to reproduce the perf gains from last week. 
From my benchmarks, searching for 5 strings in a 1M row search space:
1. `contains_warp_parallel`: 2.2ms per call => 11ms for 5 strings.
2. `multi_contains_warp_parallel`: 11.1ms for 5 strings.

It doesn't look like there's much to be gained here.

I'll keep testing a little, to see if I've missed something from last week.  If there aren't any gains to be found, I'll close this PR.

tgujar commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +373,42 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+
+  if (idx >= (num_rows * cudf::detail::warp_size * num_targets)) { return; }
+
+  auto const lane_idx   = idx % cudf::detail::warp_size;
+  auto const str_idx    = (idx / cudf::detail::warp_size) % num_rows;
+  auto const target_idx = (idx / cudf::detail::warp_size) / num_rows;
```
I think its better here to do
```
auto const str_idx    = (idx / 32) / num_targets;
auto const target_idx = (idx / 32) % num_targets;
```
Pattern is always smaller than input, so for a thread block it is always better to share input sectors than to share target sectors. I profiled this on T4 and see lesser L2 access -- lower long scoreboard stalls.

mythrocks commented:
Hat tip to @tgujar for this one:

For the multi-string kernel, [this part](https://github.com/rapidsai/cudf/pull/15536/files#diff-048f86c21559b14f64f86aaeaa57776d366c3a4948a5aba7c0ab1a3801be87bcR389-R391) is interesting:
```c++
  auto const lane_idx   = idx % cudf::detail::warp_size;
  auto const str_idx    = (idx / cudf::detail::warp_size) % num_rows;
  auto const target_idx = (idx / cudf::detail::warp_size) / num_rows;
```
I was iterating such that the first `num_rows` rows are checked for the first target string, then next `num_rows` for the second, etc.
The suggestion was to switch the order so that the first input string is checked for all `t` target strings before going to the second input string. 

On a 1M input column, with 5 search strings, calling the multi-contains kernel takes 9.4ms. Calling the single contains-kernel 5 times sequentially takes 10.64ms.  It looks like a 11% speedup.

hyperbolic2346 commented on cpp/src/strings/search/find.cu:
```diff
@@ -407,6 +444,67 @@ std::unique_ptr<column> contains_warp_parallel(strings_column_view const& input,
+                 [&](auto const& results_column) {
+                   return results_column->mutable_view().template data<bool>();
+                 });
+  host_device_results_list.host_to_device_sync(stream);
```
`rmm::exec_policy_nosync` might be useful as well.

davidwendt commented on cpp/src/strings/search/find.cu:
```diff
@@ -407,6 +444,67 @@ std::unique_ptr<column> contains_warp_parallel(strings_column_view const& input,
+      return bool_column;
+    });
+  auto results_list =
+    std::vector<std::unique_ptr<column>>(results_iter, results_iter + targets.size());
```
https://github.com/rapidsai/cudf/pull/15536#issuecomment-2058892406

mythrocks replied:
Agreed.  I have removed the call to `thrust::fill`.  I'm setting the `true/false` values directly in the kernel.

There's also a short-ckt path now, for cases where the needle's value exceeds the size of the haystack.

mythrocks commented:
On my last update, I indicated some positive news regarding total kernel times:  Batching calls to `strings::contains()` for multiple targets reduces the total time spent in the kernel. For 100M rows, 10 targets:
1. Single target kernel: 1.3s 
2. Multi-target  kernel: 0.9s

However, the end-to-end query times did not improve.  The cumulative time spent in `strings::contains()` was very high with the multi-kernel:
1. Single target kernel:  498ms
2. Multi-target  kernel: 1843ms

I've removed the call to `thrust::fill`. Per advice from @nvdbaranec and @abellina, I've switched to using `rmm::device_uvector` instead of `io::hostdevice_vector`, and turned the pinned-pool allocator. There have been improvements:

Algo | Device datastructure | Pinned Pool setting | Avg Kernel time (ms) | Total Kernel Time (ms) | Total function time (ms)
-- | -- | -- | -- | -- | --
Baseline (1 target/kernel x 10 targets) | - | 0g | 0.545 | 1091 | 498
Baseline (1 target/kernel x 10 targets) | - | 8g | 0.565 | 1131 | 562.59
Multi Kernel (10 targets/kernel) | io::hostdev_vector | 0g | 4.61 | 922.05 | 1885
Multi Kernel (10 targets/kernel) | io::hostdev_vector | 8g | 4.69 | 938.167 | 252.191
Multi Kernel (10 targets/kernel) | rmm::device_vector | 0g | 4.604 | 920.875 | 389.096
Multi Kernel (10 targets/kernel) | rmm::device_vector | 8g | 4.7 | 939.957 | 231.817

The gains in the total kernel time remain about the same:
![string_contains kernel times](https://github.com/rapidsai/cudf/assets/5607330/9578158d-4304-4dae-88ce-ef5c8b9cbdc8)

But there is reasonable improvement in the time spent in `cudf::strings::contains`:
![string_contains function times](https://github.com/rapidsai/cudf/assets/5607330/c76cec9b-c827-467a-8703-142b170a65ac)

It looks like we're closer to the solution.  It might warrant more testing with the user's actual data.

tgujar commented on cpp/src/strings/search/find.cu:
```diff
@@ -407,6 +454,69 @@ std::unique_ptr<column> contains_warp_parallel(strings_column_view const& input,
+{
+  auto const num_targets = static_cast<size_type>(targets.size());
+  CUDF_EXPECTS(not targets.empty(), "Must specify at least one target string.");
+  CUDF_EXPECTS(std::all_of(targets.begin(),
```
We end up with repeated sync because of 1 byte pageable transfers for `is_valid` call. This doesn't seem to be a bottleneck for low number of targets but could become one if we were searching for more targets. I think we could avoid this if targets were in a column instead since the null mask is a contiguous allocation
![image](https://github.com/rapidsai/cudf/assets/23545205/bded3a6b-fb3a-417e-a575-eef19ad49131)

mythrocks replied:
Fair point (although the `contains_warp_parallel` kernel might have the same challenge).  I'll hit this on the next go around.

davidwendt replied:
https://github.com/rapidsai/cudf/pull/15536#discussion_r1566415863

davidwendt commented on cpp/src/strings/search/find.cu:
```diff
@@ -372,6 +372,53 @@ CUDF_KERNEL void contains_warp_parallel_fn(column_device_view const d_strings,
+       !found && ((i + d_target.size_bytes()) <= d_str.size_bytes());
+       i += cudf::detail::warp_size) {
+    // check the target matches this part of the d_str data
+    if (d_target.compare(d_str.data() + i, d_target.size_bytes()) == 0) { found = true; }
```
I tried this in https://github.com/rapidsai/cudf/pull/15495 and it was faster for longer strings
```
| width |   rows   | hr |   Ref Time |   Cmp Time |         Diff |   %Diff |
|-------|----------|----|------------|------------|--------------|---------|
|  32   |  260000  | 20 |  93.142 us |  88.831 us |    -4.310 us |  -4.63% |
|  64   |  260000  | 20 | 203.619 us | 203.895 us |     0.276 us |   0.14% |
|  128  |  260000  | 20 | 242.877 us | 204.080 us |   -38.797 us | -15.97% |
|  256  |  260000  | 20 | 385.031 us | 286.262 us |   -98.769 us | -25.65% |
|  512  |  260000  | 20 | 667.015 us | 460.092 us |  -206.923 us | -31.02% |
| 1024  |  260000  | 20 |   1.239 ms | 805.203 us |  -433.541 us | -35.00% |
|  32   | 1953000  | 20 | 359.205 us | 359.754 us |     0.549 us |   0.15% |
|  64   | 1953000  | 20 |   1.133 ms |   1.134 ms |     1.066 us |   0.09% |
|  128  | 1953000  | 20 |   1.611 ms |   1.328 ms |  -282.885 us | -17.56% |
|  256  | 1953000  | 20 |   2.666 ms |   1.941 ms |  -724.912 us | -27.19% |
|  512  | 1953000  | 20 |   4.754 ms |   3.229 ms | -1524.398 us | -32.07% |
| 1024  | 1953000  | 20 |   8.998 ms |   5.792 ms | -3205.510 us | -35.63% |
|  32   | 16777216 | 20 |   2.747 ms |   2.747 ms |     0.334 us |   0.01% |
|  64   | 16777216 | 20 |   9.149 ms |   9.148 ms |    -1.015 us |  -0.01% |
|  32   |  260000  | 80 |  95.208 us |  94.667 us |    -0.541 us |  -0.57% |
|  64   |  260000  | 80 | 201.054 us | 202.984 us |     1.930 us |   0.96% |
|  128  |  260000  | 80 | 269.381 us | 234.155 us |   -35.225 us | -13.08% |
|  256  |  260000  | 80 | 411.775 us | 314.839 us |   -96.935 us | -23.54% |
|  512  |  260000  | 80 | 688.695 us | 489.677 us |  -199.018 us | -28.90% |
| 1024  |  260000  | 80 |   1.266 ms | 850.385 us |  -415.672 us | -32.83% |
|  32   | 1953000  | 80 | 427.888 us | 428.361 us |     0.473 us |   0.11% |
|  64   | 1953000  | 80 |   1.130 ms |   1.134 ms |     4.034 us |   0.36% |
|  128  | 1953000  | 80 |   1.801 ms |   1.557 ms |  -243.390 us | -13.52% |
|  256  | 1953000  | 80 |   2.874 ms |   2.160 ms |  -714.607 us | -24.86% |
|  512  | 1953000  | 80 |   4.918 ms |   3.445 ms | -1472.601 us | -29.94% |
| 1024  | 1953000  | 80 |   9.206 ms |   6.152 ms | -3054.411 us | -33.18% |
|  32   | 16777216 | 80 |   3.350 ms |   3.355 ms |     4.182 us |   0.12% |
|  64   | 16777216 | 80 |   9.117 ms |   9.122 ms |     5.016 us |   0.06% |

```

The `hr` is the hit-rate (i.e. 20 means the target is found in 20% of the rows)

tgujar replied:
Thanks David! Idea here is to reduce the number of request to L2

mythrocks commented:
From my experimentation, it appears that we might have taken this nearly as far as we can with this algo.  There will be incremental improvements if we allocate the multiple targets in a contiguous allocation, and space out the indices for consecutive threads.

I will try make those changes available in this PR shortly.

I've been chasing another lead with the approach detailed in [this paper](https://link.springer.com/article/10.1007/s00778-015-0409-y).  This proposes to implement KMP or Aho Corasick on CUDA.  It's out of scope for this PR, but it's a promising option.

mythrocks commented:
I've moved this over to 24.08, rather than rush this to 24.06.  There is a parallel effort on to explore KMP or Aho-Corasick, to see if that fits better.

mythrocks commented:
I've been unable to spend time on this.  

I'm closing this for now.  I'll try hit this again later, with the modified KMP approach.
